content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,172 @@
+As its name suggests, it will *stop* the function execution *if* an object does *not* pass some tests.
+
+```{r, error = TRUE}
+say_hello <- function(name) {
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+
+say_hello("Bob")
+say_hello(404)
+```
+
+However, as you can see in this example, the error message is not in plain English but contains some code instructions.
+This can hinder understanding of the issue.
+
+Because of this, `stopifnot()` was improved in R 4.0.0:
+
+> stopifnot() now allows customizing error messages via argument names, thanks to a patch proposal by Neal Fultz in PR\#17688.
+
+This means we can now provide a clearer error message directly in `stopifnot()` [^1]:
+
+[^1]: Read [the tidyverse style guide](https://style.tidyverse.org/error-messages.html) for more guidance on how to write good error messages.
+
+```{r, error = TRUE}
+say_hello <- function(name) {
+  stopifnot("`name` must be a character." = is.character(name))
+  paste("Hello", name)
+}
+
+say_hello(404)
+```
+
+This is clearly a really great improvement to the functionality of base R. However, we can see from this example that we could create the error message programmatically based on the contents of the test.
+Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
+This way, you don't have to repeat yourself which is generally a good aim [^2].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^2]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
+
+```{r}
+say_hello <- function(name) {
+  # Among other things, check_string() checks that we provide a 
+  # character object of length one
+  checkmate::assert_string(name)
+  paste("Hello", name)
+}
+```
+
+```{r, error = TRUE}
+say_hello(404)
+```
+
+```{r, error = TRUE}
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
```

Comment:
I was also wondering whether, for some packages that are not testthat, the number of revdeps could be interesting?

Reply:
Oh well, they are actually all on CRAN expected check. I might remove check (which is interesting but might have an overly complex syntax for a short demo) and add revdeps.

Reply:
I keep going back and forth with this but they all have 0 revdeps, excepted asserthat, which has 11 revdeps. I think I will not include this.